src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -27,47 +27,92 @@ GcsActorScheduler::GcsActorScheduler(
-    return;
+  if (RayConfig::instance().gcs_actor_scheduling_enabled()) {
+    ScheduleByGcs(actor);
+  } else {
+    ScheduleByRaylet(actor);
   }
+}
+
+void GcsActorScheduler::ScheduleByGcs(std::shared_ptr<GcsActor> actor) {
+  auto callback = [this, actor](const NodeID &node_id) {
+    auto node = gcs_node_manager_.GetAliveNode(node_id);
+    RAY_CHECK(node.has_value());
+
+    // Update the address of the actor as it is tied to a node.
+    rpc::Address address;
+    address.set_raylet_id(node.value()->node_id());
+    actor->UpdateAddress(address);
+
+    RAY_CHECK(node_to_actors_when_leasing_[actor->GetNodeID()]
+                  .emplace(actor->GetActorID())
+                  .second);
+
+    auto &acquired_resources = actor->GetMutableAcquiredResources();
+    acquired_resources = ResourceMapToResourceRequest(
+        actor->GetCreationTaskSpecification().GetRequiredResources().GetResourceMap(),
+        false);
+    // Lease worker directly from the node.
+    LeaseWorkerFromNode(actor, node.value());
+  };
+
+  // Queue and schedule the actor locally (gcs).
+  cluster_task_manager_->QueueAndScheduleTask(actor->GetCreationTaskSpecification(),
+                                              /*forward_to=*/NodeID::Nil(),
+                                              /*use_required_resources=*/false,
+                                              callback);
+}
 
-  // Update the address of the actor as it is tied to a node.
-  rpc::Address address;
-  address.set_raylet_id(node.value()->node_id());
-  actor->UpdateAddress(address);
+void GcsActorScheduler::ScheduleByRaylet(std::shared_ptr<GcsActor> actor) {
+  auto callback = [this, actor](const NodeID &node_id) {
+    auto node = gcs_node_manager_.GetAliveNode(node_id);
+    RAY_CHECK(node.has_value());
 
-  RAY_CHECK(node_to_actors_when_leasing_[actor->GetNodeID()]
-                .emplace(actor->GetActorID())
-                .second);
+    // Update the address of the actor as it is tied to a node.
+    rpc::Address address;
+    address.set_raylet_id(node.value()->node_id());
+    actor->UpdateAddress(address);
 
-  // Lease worker directly from the node.
-  LeaseWorkerFromNode(actor, node.value());
+    RAY_CHECK(node_to_actors_when_leasing_[actor->GetNodeID()]
+                  .emplace(actor->GetActorID())
+                  .second);
+
+    // Lease worker directly from the node.
+    LeaseWorkerFromNode(actor, node.value());
+  };
+
+  // If an actor has resource requirements, we will try to forward it to the owner's node
+  // for scheduling. Otherwise, it would be forwarded to a random remote node for
+  // scheduling.
+  const auto task_spec = actor->GetCreationTaskSpecification();
+  NodeID node_id = task_spec.GetRequiredResources().IsEmpty() ? NodeID::Nil()
+                                                              : actor->GetOwnerNodeID();
+  cluster_task_manager_->QueueAndScheduleTask(task_spec,
```

Comment:
this seems wrong. In this case I think we should directly talk to remote cluster_task_manager, instead of adding this `forward_to` logic to cluster task manager.  Is there any reason we can't do that?